pinot-core/src/main/java/org/apache/pinot/core/data/table/OrderedIndexedTableResizer.java

```diff
@@ -0,0 +1,260 @@
+   */
+  @VisibleForTesting
+  IntermediateRecord getIntermediateRecord(Record record) {
+    Comparable[] intermediateRecordValues = new Comparable[_numOrderBy];
+
+    Object[] keyColumns = record.getKey().getColumns();
+    Object[] aggregationColumns = record.getValues();
+    for (int i = 0; i < _numOrderBy; i++) {
+      OrderByInfo orderByInfo = _orderByInfos.get(i);
+      Comparable comparable;
+      if (orderByInfo.getOrderByType().equals(OrderByType.KEY_COLUMN)) {
+        Object keyColumn = keyColumns[orderByInfo.getIndex()];
+        comparable = (Comparable) keyColumn; // FIXME: is this the right way to get Comparable? will it work for BYTES?
+      } else {
+        Object aggregationColumn = aggregationColumns[orderByInfo.getIndex()];
+        AggregationFunction aggregationFunction = orderByInfo.getAggregationFunction();
+        if (!aggregationFunction.isIntermediateResultComparable()) {
+          comparable = aggregationFunction.extractFinalResult(aggregationColumn);
+        } else {
+          comparable = (Comparable) aggregationColumn;
+        }
+      }
+      intermediateRecordValues[i] = comparable;
+    }
+    return new IntermediateRecord(record.getKey(), intermediateRecordValues);
+  }
+
+  /**
+   * Trim recordsMap to trimToSize, based on order by information
+   */
+  @Override
+  void resizeRecordsMap(Map<Key, Record> recordsMap, int trimToSize) {
+
+    int numRecordsToEvict = recordsMap.size() - trimToSize;
+
+    if (numRecordsToEvict > 0) {
+      int size;
+      Comparator<IntermediateRecord> comparator;
+
+      if (numRecordsToEvict < trimToSize) { // num records to evict is smaller than num records to retain
+        size = numRecordsToEvict;
+        // make PQ of records to evict
+        comparator = _intermediateRecordComparator;
+      } else { // num records to retain is smaller than num records to evict
+        size = trimToSize;
+        // make PQ of records to retain
+        comparator = _intermediateRecordComparator.reversed();
+      }
+
+      PriorityQueue<IntermediateRecord> priorityQueue =
+          new PriorityQueue<>(size, comparator);
+
+      for (Record record : recordsMap.values()) {
+
+        IntermediateRecord intermediateRecord = getIntermediateRecord(record);
+        if (priorityQueue.size() < size) {
+          priorityQueue.offer(intermediateRecord);
+        } else {
+          IntermediateRecord peek = priorityQueue.peek();
+          if (priorityQueue.comparator().compare(peek, intermediateRecord) < 0) {
+            priorityQueue.poll();
+            priorityQueue.offer(intermediateRecord);
+          }
+        }
+      }
+
+      if (numRecordsToEvict < trimToSize) { // PQ contains records to evict
+        for (IntermediateRecord evictRecord : priorityQueue) {
+          recordsMap.remove(evictRecord.getKey());
+        }
+      } else { // PQ contains records to retain
+        Set<Key> keysToRetain = new HashSet<>(priorityQueue.size());
```

Comment:
This will almost always cause a resize. You can use ObjectOpenHashSet to avoid this, or take the load factor into the initial capacity